ensime_shared/ensime.py

```diff
@@ -584,9 +584,13 @@ def handle_completion_info_list(self, call_id, payload):
         """Handler for a completion response."""
         completions = payload["completions"]
         self.log("handle_completion_info_list: in")
-        self.suggestions = [completion_to_suggest(c) for c in completions]
+        self.suggestions = [completion_to_suggest(c, self.format_type_params()) for c in completions]
         self.log("handle_completion_info_list: {}".format(self.suggestions))
 
+    def format_type_params(self):
+        """NOP as it's not available for Server 1.0"""
+        return lambda _: ""
+
```

Comment:
@ches I agree mostly on everything. 

Strongly agree on improving testability, when I implemented completion formatting, the first thing I did was to refactor methods into just module functions. Having to create a full client just to test a bunch of functions that can be tested in isolation is a big stopper for adding new tests.
Not sure about the abstract class for the handlers, but I guess that having the interface defined in a single place is something good, so kinda agree on that too.
Don't agree on the runtime mixin bit, I was taking a look and it seems that you need to use that `__metaclass__` thing, which in my opinion complicates the code for not clear win. We can instead have fixed mixins for v1 and v2 and instantiate one of them based on the `server2` flag.

I can simplify even more the type params formatting, moving the __NOP__ inside `completion_to_suggest` and turn it on based on the server version flag, no need for overriding any method for this very case, so it is a bad example for testing the code branching.
I think I'll close this PR and open one for the mixin refactor + server 2.0 support, and a simple one just for the type params formatting.

To recap, I'll factor out just the handlers to a separate set of mixins, let's say `handler.py` with the abstract `ProtocolHandler` class and the concrete ones for V1 and V2, I think we can keep them all in the same file for now.

What the others think about it?